File: frontend/packages/api-client/meal_plan_tasks.ts
import { Axios } from 'axios';
import format from 'string-format';

import { APIResponse, MealPlanTask, MealPlanTaskStatusChangeRequestInput } from '@dinnerdonebetter/models';

import { backendRoutes } from './routes';

export async function getMealPlanTask(
  client: Axios,
  mealPlanID: string,
  mealPlanTaskID: string,
): Promise<MealPlanTask> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<MealPlanTask>>(
      format(backendRoutes.MEAL_PLAN_TASKS, mealPlanID, mealPlanTaskID),
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getMealPlanTasks(client: Axios, mealPlanID: string): Promise<MealPlanTask[]> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<MealPlanTask[]>>(format(backendRoutes.MEAL_PLAN_TASKS, mealPlanID));

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function updateMealPlanTaskStatus(
  client: Axios,
  mealPlanID: string,
  mealPlanTaskID: string,
  input: MealPlanTaskStatusChangeRequestInput,
): Promise<MealPlanTask> {
  return new Promise(async function (resolve, reject) {
    const response = await client.patch<APIResponse<MealPlanTask>>(
      format(backendRoutes.MEAL_PLAN_TASK, mealPlanID, mealPlanTaskID),
      input,
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

File: backend/pkg/apiclient/meal_plan_tasks.go
package apiclient

import (
	"context"

	"github.com/dinnerdonebetter/backend/internal/observability"
	"github.com/dinnerdonebetter/backend/internal/observability/keys"
	"github.com/dinnerdonebetter/backend/internal/observability/tracing"
	"github.com/dinnerdonebetter/backend/pkg/apiclient/generated"
	"github.com/dinnerdonebetter/backend/pkg/types"
)

// GetMealPlanTask gets a meal plan task.
func (c *Client) GetMealPlanTask(ctx context.Context, mealPlanID, mealPlanTaskID string) (*types.MealPlanTask, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealPlanID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	if mealPlanTaskID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanTaskIDKey, mealPlanTaskID)
	tracing.AttachToSpan(span, keys.MealPlanTaskIDKey, mealPlanTaskID)

	res, err := c.authedGeneratedClient.GetMealPlanTask(ctx, mealPlanID, mealPlanTaskID)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "get meal plan task")
	}

	var apiResponse *types.APIResponse[*types.MealPlanTask]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meal plan task")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// CreateMealPlanTask creates a meal plan task.
func (c *Client) CreateMealPlanTask(ctx context.Context, mealPlanID string, input *types.MealPlanTaskCreationRequestInput) (*types.MealPlanTask, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealPlanID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	if input == nil {
		return nil, ErrInvalidIDProvided
	}

	body := generated.CreateMealPlanTaskJSONRequestBody{}
	c.copyType(&body, input)

	res, err := c.authedGeneratedClient.CreateMealPlanTask(ctx, mealPlanID, body)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "get meal plan task")
	}

	var apiResponse *types.APIResponse[*types.MealPlanTask]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meal plan task")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// UpdateMealPlanTaskStatus updates a meal plan task.
func (c *Client) UpdateMealPlanTaskStatus(ctx context.Context, mealPlanID string, input *types.MealPlanTaskStatusChangeRequestInput) (*types.MealPlanTask, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if input == nil {
		return nil, ErrNilInputProvided
	}

	if mealPlanID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	if err := input.ValidateWithContext(ctx); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "validating input")
	}

	body := generated.UpdateMealPlanTaskStatusJSONRequestBody{}
	c.copyType(&body, input)

	res, err := c.authedGeneratedClient.UpdateMealPlanTaskStatus(ctx, mealPlanID, input.ID, body)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "create meal plan task")
	}

	var apiResponse *types.APIResponse[*types.MealPlanTask]
	if err = c.unmarshalBody(ctx, res, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "creating meal plan task")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}
